tests/ga/test_remoteaccess_handler.py

```diff
@@ -0,0 +1,141 @@
+# Copyright Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Requires Python 2.6+ and Openssl 1.0+
+#
+
+import base64
+import errno as errno
+import glob
+import random
+import string
+import tempfile
+import uuid
+
+from datetime import datetime, timedelta, tzinfo
+from azurelinuxagent.common.protocol import get_protocol_util
+import azurelinuxagent.common.utils.shellutil as shellutil
+from azurelinuxagent.common.protocol.wire import *
+from azurelinuxagent.common.osutil import get_osutil
+from azurelinuxagent.common.osutil.default import DefaultOSUtil
+from azurelinuxagent.common.utils.cryptutil import CryptUtil
+from azurelinuxagent.common.exception import CryptError
+from azurelinuxagent.ga.remoteaccess import RemoteAccessHandler, REMOTE_USR_EXPIRATION_FORMAT, MAX_TRY_ATTEMPT
+from tests.common.osutil.mock_osutil import MockOSUtil
+from tests.protocol.mockwiredata import WireProtocolData, DATA_FILE
+from tests.tools import *
+
+class TestRemoteAccessHandler(AgentTestCase):
+    
+    def test_add_user(self):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "abc@123"
+        tstuser = "foobar"
+        expiration = datetime.utcnow() + timedelta(days=1)
+        pwd = self.encrypt_string(tstpassword)
+        rah.add_user(tstuser, pwd, expiration)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertTrue(tstuser in os_util.all_users)
+        actual_user = os_util.all_users[tstuser]
+        self.assertEqual(actual_user[1], tstpassword)
+        self.assertEqual(actual_user[7], (expiration - datetime.utcnow() + timedelta(days=1)).days)
+        self.assertEqual(actual_user[4], "JIT Account")
+            
+    def test_delete_user(self):
+        rah = RemoteAccessHandler()
+        rah.os_util = MockOSUtil()
+        tstpassword = "abc@123"
+        tstuser = "foobar"
+        expiration = datetime.utcnow() + timedelta(days=1)
+        pwd = self.encrypt_string(tstpassword)
+        rah.add_user(tstuser, pwd, expiration)
+        os_util = rah.os_util
+        os_util.__class__ = MockOSUtil
+        self.assertTrue(tstuser in os_util.all_users)
+        rah.delete_user(tstuser)
+        self.assertFalse(tstuser in os_util.all_users)
+    
+    def test_handle_new_user(self):
```

Comment:
Missing tests for handling (via `handle_remote_access()`) an existing user

- Subcase 1: existing user not yet expired
- Subcase 2: existing user that has already expired

Reply:
yes, I have quite a few more tests to implement here.  I will make sure these are included.

---
In reply to: [192188495](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r192188495) [](ancestors = 192188495)

Reply:
added tests.  Note, the only logic is if the account already exists or not.  We don't extend expiration dates in JIT.  The user would just submit a new request and get a new userid.

---
In reply to: [192224617](https://github.com/Azure/WALinuxAgent/pull/1171#discussion_r192224617) [](ancestors = 192224617,192188495)